docs/rfcs/2025-04-14-storage-keys.md

```diff
@@ -0,0 +1,246 @@
+
+# Storage Encryption Key Management
+
+## Summary
+
+As a precursor to adding new encryption capabilities to Neon's storage services, this RFC proposes
+mechanisms for creating and storing fine-grained encryption keys for user data in Neon.  We aim
+to provide at least tenant granularity, but will use timeline granularity when it is simpler to do
+so.
+
+Out of scope:
+- We describe lifecycle of keys here but not the encryption of user data with these keys.
+- We describe an abstract KMS interface, but not particular platform implementations (such as how
+  to authenticate with KMS).
+
+## Terminology
+
+_wrapped/unwrapped_: a wrapped encryption key is a key encrypted by another key.  For example, the key for
+encrypting a timeline's pageserver data might be wrapped by some "root" key for the tenant's user account, stored in a KMS system.
+
+_key hierarchy_: the relationships between keys which wrap each other. For example, a layer file key might
+be wrapped by a pageserver timeline key, which is wrapped by a tenant's root key.
+
+## Design Choices
+
+Storage: S3 will be the store of record for wrapped keys
+
+Separate keys: Safekeeper and Pageserver will use independent keys.
+
+AES256: rather than building a generic system for keys, we will assume that all the keys
+we manage are AES256 keys -- this is the de-facto standard for enterprise data storage.
+
+Per-object keys: rather than encrypting data objects (layer files and segment files) with
+the tenant keys directly, they will be encrypted with separate keys.  This avoids cryptographic
+safety issues from re-using the same key for large quantities of potentially repetitive plaintext.
+
+Key storage is optional at a per-tenant granularity: eventually this would be on by default, but:
+- initially only some environments will have a KMS set up.
+- Encryption has some overhead and it may be that some tenants don't want or need it.
+
+## Design
+
+### Summary of format changes
+
+- Pageserver layer files and safekeeper segment objects get new metadata fields to
+  store wrapped key and version of the wrapping key
```

Comment:
Storing inline makes the object more self-contained.
But on the other hand we have no good format evolution story for layer files (I think).
We could sidestep format evolution by storing the key in index part.
That also avoids changing the format for this "Milestone 1" target where we only encrypt in S3.

Reply:
Hm, reading later, I see you propose to store the key version in layer file metadata. 
So I guess by "gets new metadata fields" you mean "... in index_part.json".

Reply:
Clarified during kick-off call:

- primary location for the wrapped key + wrapping key version is in index part
- backup also is in object metadata tags (to make object more free-standing)